core/src/main/scala/akka/persistence/jdbc/state/scaladsl/DurableStateSequenceActor.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2014 - 2019 Dennis Vriend <https://github.com/dnvriend>
+ * Copyright (C) 2019 - 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.jdbc.state.scaladsl
+
+import akka.actor.{ Actor, ActorLogging, Props, Status, Timers }
+import akka.pattern.pipe
+import akka.persistence.jdbc.config.DurableStateSequenceRetrievalConfig
+import akka.stream.Materializer
+import akka.stream.scaladsl.Sink
+
+import scala.collection.immutable.NumericRange
+import scala.concurrent.duration.FiniteDuration
+
+object DurableStateSequenceActor {
+  def props[A](stateStore: JdbcDurableStateStore[A], config: DurableStateSequenceRetrievalConfig)(
+      implicit materializer: Materializer): Props = Props(new DurableStateSequenceActor(stateStore, config))
+
+  private case object QueryOrderingIds
+  private case class NewOrderingIds(originalOffset: Long, elements: Seq[OrderingId])
+
+  private case class ScheduleAssumeMaxOrderingId(max: OrderingId)
+  private case class AssumeMaxOrderingId(max: OrderingId)
+
+  case object GetMaxOrderingId
+  case class MaxOrderingId(maxOrdering: OrderingId)
+
+  private case object QueryOrderingIdsTimerKey
+  private case object AssumeMaxOrderingIdTimerKey
+
+  private type OrderingId = Long
+
+  /**
+   * Efficient representation of missing elements using NumericRanges.
+   * It can be seen as a collection of OrderingIds
+   */
+  private case class MissingElements(elements: Seq[NumericRange[OrderingId]]) {
+    def addRange(from: OrderingId, until: OrderingId): MissingElements = {
+      val newRange = from.until(until)
+      MissingElements(elements :+ newRange)
+    }
+    def contains(id: OrderingId): Boolean = elements.exists(_.containsTyped(id))
+    def isEmpty: Boolean = elements.forall(_.isEmpty)
+  }
+  private object MissingElements {
+    def empty: MissingElements = MissingElements(Vector.empty)
+  }
+}
+
+/**
+ * To support the changesByTag query, this actor keeps track of which rows are visible in the database.
+ * This is required to guarantee the changesByTag does not skip any rows in case rows with a higher (ordering) id are
+ * visible in the database before rows with a lower (ordering) id.
+ */
+class DurableStateSequenceActor[A](stateStore: JdbcDurableStateStore[A], config: DurableStateSequenceRetrievalConfig)(
```

Comment:
I don't think this will do the trick for the durable state. The `JournalSequenceActor` was designed to detect gaps that occur when a more recent ordering (`n`) shows up on the table before  a previous one (`n-1`). 

Although that could also happen for durable state, we should have something that can also detect missing updates. 

If we have:

```scala
1 | abc
2 | def
```

And both get updated, `abc` becomes 3 and `def` becomes 4. It's possible `def` transaction commits first and the query see the table as: 

```scala
1 | abc
4 | def
```

The stream delives 4, user saves the offset and now 3 will never be delivered. We lost an update.

Reply:
@octonato let me first clear the confusion in terminology. What u r saying as `ordering` for the `JournalSequenceActor` it is the `global_offset` in `DurableStateSequenceActor` - am I correct ? So in the implementation of the actor, `OrderingId` refers to `ordering` for the journal and the `offset` for the state implementation. `JournalSequenceActor` detects missing `ordering`, while `DurableStateSequenceActor` detects missing `offsets`. Then why should we have a lost update in this implementation ?

Reply:
@octonato In other words why should the actor not detect that 3 is missing and deliver it as well ?

Reply:
> JournalSequenceActor detects missing ordering, while DurableStateSequenceActor detects missing offsets

That is right.

I assume that Renato's example is using global_offset as the first column and persistence_id as the second column.

For that example, the DurableStateSequenceActor will detect that offset 3 is missing when it sees 4, it will query again, and will eventually find 3. 4 is not delivered until 3 has been seen.

However, there is a another scenario where we have a problem. If there are several quick updates for the same persistence_id we will see the intermediate updates as missing offsets. Those will be retried and give up on those after timeout. All good in the end, but it will drastically slow down the queries.

```
global_offset | persistence_id | revision
---------------------------------------
17            | abc            | 1
```
2 quick updates of "abc" and the query will see
```
global_offset | persistence_id | revision
---------------------------------------
19            | abc            | 3
```
missing offset 18

could be more complex with several persistence_ids updated quickly
```
global_offset | persistence_id | revision
---------------------------------------
17            | abc            | 1
18            | def            | 1
```
2 quick updates of abc and 2 of def
```
global_offset | persistence_id | revision
---------------------------------------
20            | abc            | 3
22            | def            | 3
```
missing offset 19 and 21

One way to solve that could be to also track seen persistence_id/revision. In above example, we could then see that a total of 2 revisions are missing and that matches 2 missing offsets. Then it should be safe to say that we have seen all changes.

Reply:
@patriknw perhaps I am missing something. When we query for the `changes` we want to get the latest offset for a `persistence_id` for some specified `tag` and _NOT_ all the intermediate ones. Then what is the harm in the above case if we do not see offset 19 and 21 ?

Reply:
We can't know if 19 and 21 were written by abc and def respectively. They could have been written by xyz and then we can't ignore them.

Reply:
@patriknw But if they are of a different persistence_id, then the actor will eventually fetch them in the query. Maybe I am missing how can the actor miss them .. If 19 and 21 are from diff persistence_ids then they will be inserted and the query will get them. If they are from the same persistence id, then also the actor will get at least the max of them.

Reply:
That is right, with current implementation, but my point was:

> Those will be retried and give up on those after timeout. All good in the end, but it will drastically slow down the queries.

We will always see such timeout delays when there are quick updates to the same persistence_id. Since that seems like a rather common scenario it would be good to find a solution where we can avoid these delays (in most cases).

Reply:
ah .. ok .. makes sense now.